gusto/terms.py

```diff
@@ -0,0 +1,43 @@
+from abc import ABCMeta, abstractmethod
+from firedrake import FacetNormal, div, dx, inner
+
+
+class Term(object, metaclass=ABCMeta):
+
+    off_centering = 0.5
+
+    def __init__(self, state, test):
+        self.state = state
+        self.test = test
+        self.parameters = state.parameters
+        self.n = FacetNormal(state.mesh)
+
+    @abstractmethod
+    def evaluate(self, q, fields):
+        pass
+
+    def __call__(self, q, fields):
+        return self.evaluate(q, fields)
+
+
+class ShallowWaterPressureGradientTerm(Term):
+
+    def evaluate(self, q, fields):
+        g = self.parameters.g
+        D = fields("D")
+        return g*div(self.test)*D*dx
```

Comment:
There's something funny in the design here. Somehow the `Term` knows which coefficients to extract from the state but doesn't know which test function. This causes a kind of broken concept of who owns what information.

Reply:
Yes, I think the parameters should be part of the equation class, rather than state.

Reply:
Sorry, I misspoke. I mean which function space it's taking pressure gradient against. Same issue.